demosplan/DemosPlanCoreBundle/Addon/Registrator.php

```diff
@@ -0,0 +1,101 @@
+use demosplan\DemosPlanCoreBundle\Utilities\DemosPlanPath;
+use Exception;
+use Symfony\Component\Yaml\Yaml;
+
+final class Registrator
+{
+    public const ADDON_DIRECTORY = '/addons/';
+    public const ADDON_CACHE_DIRECTORY = '/addons/cache/';
+
+    private const ADDON_YAML_INLINE_DEPTH = 100;
+
+    private PackageInformation $packageInformation;
+
+    private array $addons;
+
+    public function __construct()
+    {
+        $this->packageInformation = new PackageInformation();
+        $this->addons = AddonManifestCollection::load();
+    }
+
+    /**
+     * Checks if a given composer definition is a correct representation of an addon.
+     * If so and if that addon is not yet installed, it will be added to the list of installed addons.
+     */
+    public function register(PackageInterface $addonComposerDefinition): string
+    {
+        if (PackageInformation::ADDON_COMPOSER_TYPE !== $addonComposerDefinition->getType()) {
+            throw AddonException::invalidType($addonComposerDefinition->getName(), $addonComposerDefinition->getType());
+        }
+
+        // if addon is not in registry, then add it
+        if (!$this->isRegistered($addonComposerDefinition->getName())) {
+            $this->packageInformation->reloadPackages();
+            $this->doRegister($addonComposerDefinition);
+        }
+
+        $this->refreshAddonsYaml();
+
+        return $addonComposerDefinition->getName();
+    }
+
+    /**
+     * Writes the current collection of addons back into the addons.yaml.
+     */
+    private function refreshAddonsYaml(): void
+    {
+        $yamlContent = [
+            'addons' => $this->addons,
+        ];
+
+        file_put_contents(
+            DemosPlanPath::getRootPath('addons/addons.yaml'),
+            Yaml::dump($yamlContent, self::ADDON_YAML_INLINE_DEPTH)
+        );
+    }
+
+    private function isRegistered(string $addonName): bool
+    {
+        return array_key_exists($addonName, $this->addons);
+    }
+
+    private function doRegister(PackageInterface $addonComposerDefinition): void
+    {
+        $addonName = $addonComposerDefinition->getName();
+
+        $this->addons[$addonName] = [
+            'enabled'      => false,
+            'installed_at' => Carbon::now()->toIso8601String(),
+            'install_path' => realpath($this->packageInformation->getInstallPath($addonName)),
+            'manifest'     => $this->loadManifest($addonName),
+        ];
+    }
+
+    /**
+     * @return array<string, array<string, string>>
+     */
+    private function loadManifest(string $addonName): array
+    {
+        try {
+            // TODO: Fix manifest parsing
+            /*$config = Yaml::parseFile($this->installedAddons->getManifestPath($addonName));
+            $processor = new Processor();
+            $parsed = $processor->processConfiguration(new ManifestConfiguration(), [$config['demosplan_addon']]);
+            var_export($parsed);*/
```

Comment:
We should either try to fix it now or remove it as it is not strictly necessary

Reply:
added in [a6f0cfa](https://github.com/demos-europe/demosplan-core/pull/640/commits/a6f0cfad7075e4deeb474b49a44f62ee83ab780e)